core/src/nucleus/actions/call_zome_function.rs

```diff
@@ -201,19 +307,252 @@ impl Future for CallResultFuture {
-        let cap_call = CapabilityCall::new(Address::from(""), None);
-        assert!(!is_token_the_agent(context, &Some(cap_call)));
+
+        let mut cap_functions = CapFunctions::new();
+        cap_functions.insert("test_zome".to_string(), vec![String::from("test")]);
+        let grant =
+            CapTokenGrant::create(CapabilityType::Transferable, None, cap_functions).unwrap();
+        let grant_entry = Entry::CapTokenGrant(grant.clone());
+        let grant_addr = block_on(author_entry(&grant_entry, None, &context)).unwrap();
+        let maybe_grant = get_grant(&context, &grant_addr);
+        assert_eq!(maybe_grant, Some(grant));
+    }
+
+    #[test]
+    fn test_verify_grant() {
+        let context = test_context("alice", None);
+        let test_address1 = Address::from("agent 1");
+        let test_address2 = Address::from("some other identity");
+
+        fn zome_call_valid(context: Arc<Context>, token: &Address, addr: &Address) -> ZomeFnCall {
+            ZomeFnCall::new(
+                "test_zome",
+                make_cap_request_for_call(
+                    context.clone(),
+                    token.clone(),
+                    addr.clone(),
+                    "test",
+                    "{}",
+                ),
+                "test",
+                "{}",
+            )
+        }
+
+        let zome_call_from_addr1_bad_token = &ZomeFnCall::new(
+            "test_zome",
+            make_cap_request_for_call(
+                context.clone(),
+                Address::from("bad token"),
+                test_address1.clone(),
+                "test",
+                "{}",
+            ),
+            "test",
+            "{}",
+        );
+
+        let mut cap_functions = CapFunctions::new();
+        cap_functions.insert("test_zome".to_string(), vec![String::from("test")]);
+
+        let grant = CapTokenGrant::create(CapabilityType::Public, None, cap_functions).unwrap();
+        let token = grant.token();
+        assert!(verify_grant(
+            context.clone(),
+            &grant,
+            &zome_call_valid(context.clone(), &token, &test_address1)
+        ));
+        assert!(!verify_grant(
+            context.clone(),
+            &grant,
+            &zome_call_from_addr1_bad_token
+        ));
+
+        let mut cap_functions = CapFunctions::new();
+        cap_functions.insert("test_zome".to_string(), vec![String::from("other_fn")]);
+        let grant_for_other_fn =
+            CapTokenGrant::create(CapabilityType::Transferable, None, cap_functions).unwrap();
+        assert!(!verify_grant(
+            context.clone(),
+            &grant_for_other_fn,
+            &zome_call_valid(context.clone(), &grant_for_other_fn.token(), &test_address1)
+        ));
+
+        let mut cap_functions = CapFunctions::new();
+        cap_functions.insert("test_zome".to_string(), vec![String::from("test")]);
+        let grant =
```

Comment:
Should we split the below into a different test function since it's a different case?